### PR TITLE
Extend the C++ module scanner to handle Fortran, too.

### PR DESCRIPTION
**Pull request by jpakkane**
This will probably fail to the tests, but let's just get this out there for people to test and comment on. So ping all Fortran users, such as @scivision.

The main advantage of this approach is that we can drop all the prescan hacks we have and instead do Fortran automatically and reliably out of the box. I.e. moving module definitions from one file to another should just work without needing to run Meson reconfiguration.

### CODE REVIEW

File: mesonbuild/scripts/depscan.py [hunk 1]
```diff
@@ -15,12 +15,23 @@
...
+cpp_import_re = re.compile('\w*import ([a-zA-Z0-9]+);')
+cpp_export_re = re.compile('\w*export module ([a-zA-Z0-9]+);')
+
+FORTRAN_INCLUDE_PAT = r"^\s*#?include\s*['\"](\w+\.\w+)['\"]"
+FORTRAN_MODULE_PAT = r"^\s*\bmodule\b\s+(\w+)\s*(?:!+.*)*$"
+FORTRAN_SUBMOD_PAT = r"^\s*\bsubmodule\b\s*\((\w+:?\w+)\)\s*(\w+)"
+FORTRAN_USE_PAT = r"^\s*use,?\s*(?:non_intrinsic)?\s*(?:::)?\s*(\w+)"
+
+fortran_module_re = re.compile(FORTRAN_MODULE_PAT)
+fortran_use_re = re.compile(FORTRAN_USE_PAT)
```
Comment (dcbaker) [inline mesonbuild/scripts/depscan.py:34]: These should really all be all caps per PEP8, as they're module constants.

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -907,10 +909,15 @@ def select_sources_to_scan(self, compiled_sources):
         # in practice pick up C++ and Fortran files. If some other language
         # requires scanning (possibly Java to deal with inner class files)
         # then add them here.
+        all_suffixes = set()
+        for s in compilers.lang_suffixes['cpp']:
+            all_suffixes.add(s)
+        for s in compilers.lang_suffixes['fortran']:
+            all_suffixes.add(s)
```
Comment (dcbaker) [inline mesonbuild/backend/ninjabackend.py:916]:
```suggestion
        all_suffixes = set(compilers.lang_suffixes['cpp']) | set(compilers.lang_suffixes['fortran'])
```

**Review by dcbaker — CHANGES_REQUESTED** [review]
I have a few small nits, but overall this looks good to me. Probably should get scivision to give his input though.

File: mesonbuild/scripts/depscan.py [hunk 3]
```diff
@@ -34,9 +45,19 @@ def __init__(self, pickle_file: str, outfile: str, sources: T.List[str]):
...
-        for line in pathlib.Path(fname).read_text().split('\n'):
-            import_match = import_re.match(line)
-            export_match = export_re.match(line)
+        suffix = os.path.splitext(fname)[1][1:]
+        if suffix in lang_suffixes['fortran']:
+            self.scan_fortran_file(fname)
+        elif suffix in lang_suffixes['cpp']:
+            self.scan_fortran_file(fname)
+        else:
+            sys.exit('Can not scan files with suffix .{}.'.format(suffix))
```
Comment (scivision) [inline mesonbuild/scripts/depscan.py:54]: I think `raise SystemExit` is preferred to `sys.exit`.  However, in this case should it rather be `raise RuntimeError`?

File: mesonbuild/scripts/depscan.py [hunk 4]
```diff
@@ -15,12 +15,23 @@
...
 
 from ..backend.ninjabackend import TargetDependencyScannerInfo
+from ..compilers.compilers import lang_suffixes
 
-import_re = re.compile('\w*import ([a-zA-Z0-9]+);')
-export_re = re.compile('\w*export module ([a-zA-Z0-9]+);')
+cpp_import_re = re.compile('\w*import ([a-zA-Z0-9]+);')
+cpp_export_re = re.compile('\w*export module ([a-zA-Z0-9]+);')
+
+FORTRAN_INCLUDE_PAT = r"^\s*#?include\s*['\"](\w+\.\w+)['\"]"
```
Comment (scivision) [inline mesonbuild/scripts/depscan.py:28]: This could be in a separate PR, but to speed this up, the `#include` syntax isn't used by any compiler I've found. I added this when I was overhauling the regex for Fortran, but since realized no current compiler allows this syntax. 

Suggest:
```python
FORTRAN_INCLUDE_PAT = r"^\s*include\s*['\"](\w+\.\w+)['\"]"
```

**Comment by scivision** [comment]
Most of the Fortran project tests fail with this setup (only one I tried):

* Ninja 1.10.2
* GCC 10.2.0 (MSYS2)
* Windows 10

Typical build-time error:

```
ninja: build stopped: dependency cycle: use_syntax.exe.p/circle.mod -> use_syntax.exe.p/circle.mod.
```

**Comment by scivision** [comment]
3cd6586 fails at a list | list I think maybe it's meant to have those variables be sets?

**Comment by jpakkane** [comment]
Finally it passes all project tests. I'll fix the mypy issues later.

**Comment by scivision** [comment]
Yes this works for me with Gfortran 10 as well. 

I can't effectively test with the Intel Windows compiler yet, as the Intel oneAPI "gold" official release a few weeks ago actually broke a bunch of things that previously worked throughout the beta period, including independently breaking Meson and CMake.

https://community.intel.com/t5/Intel-oneAPI-HPC-Toolkit/Gold-oneAPI-release-broken-but-betas-worked-fine-Linux-Windows/m-p/1236017#M7421

Intel has confirmed the issues and referred to internal team.

**Comment by jpakkane** [comment]
That is unrelated to this issue, so merging.

**Comment by scivision** [comment]
yes that's correct, thanks
